src/simulations/functional_inversions/functional_inversion_utils.jl

```diff
@@ -266,132 +293,232 @@ end
+        # For the moment there is a bug when we use callbacks with SciMLSensitivity for the gradient computation
+        mb_action! = let model = container.simulation.model, cache = container.simulation.cache, glacier = glacier, step = params.solver.step
+            function (integrator)
+                if params.simulation.use_MB
+                    # Compute mass balance
+                    MB_timestep!(cache, model, glacier, step, integrator.t)
+                    apply_MB_mask!(integrator.u, cache.iceflow)
+                end
             end
         end
+        PeriodicCallback(mb_action!, params.solver.step; initial_affect=false)
+    else
+        CallbackSet()
     end
-    cb_MB = DiscreteCallback(stop_condition, mb_action!)
 
     # Create iceflow law callback
-    cb_iceflow = Huginn.build_callback(simulation.model.iceflow, simulation.cache.iceflow, simulation.cache.iceflow.glacier_idx, θ)
+    cb_iceflow = Huginn.build_callback(
+        container.simulation.model.iceflow,
+        container.simulation.cache.iceflow,
+        container.simulation.cache.iceflow.glacier_idx,
+        container.θ)
 
     cb = CallbackSet(cb_MB, cb_iceflow)
 
     # Run iceflow UDE for this glacier
-    iceflow_sol = simulate_iceflow_UDE!(θ, simulation, cb, glacier_idx)
+    iceflow_sol = simulate_iceflow_UDE!(container, cb, iceflow_prob)
 
-    # Update simulation results
-    result = Sleipnir.create_results(
-        simulation, glacier_idx, iceflow_sol, nothing;
-        tstops = simulation.parameters.solver.tstops,
-        light = !simulation.parameters.solver.save_everystep,
+    # Compute simulation results
+    return Sleipnir.create_results(
+        container.simulation, glacier_idx, iceflow_sol, nothing;
+        light = !container.simulation.parameters.solver.save_everystep,
         processVelocity = Huginn.V_from_H
     )
-
-    return result
 end
 
 
 """
     simulate_iceflow_UDE!(
-        θ,
-        simulation::SIM,
+        container::FunctionalInversionBinder,
         cb::SciMLBase.DECallback,
-        glacier_idx::I,
-    ) where {I <: Integer, SIM <: Simulation}
+        iceflow_prob::ODEProblem,
+    )
 
 Make forward simulation of the iceflow UDE.
 """
 function simulate_iceflow_UDE!(
-    θ,
-    simulation::SIM,
+    container::FunctionalInversionBinder,
     cb::SciMLBase.DECallback,
-    glacier_idx::I,
-) where {I <: Integer, SIM <: Simulation}
-
-    model = simulation.model
-    cache = simulation.cache
-    params = simulation.parameters
-
-    # Define closure with apply_parametrization inside the function call
-    SIA2D_UDE_closure(H, θ, t) = SIA2D_UDE(H, θ, t, simulation)
-
-    iceflow_prob = ODEProblem(
-        SIA2D_UDE_closure,
-        cache.iceflow.H₀,
-        params.simulation.tspan,
-        θ;
-        tstops=params.solver.tstops,
-    )
-
+    iceflow_prob::ODEProblem,
+)
+    params = container.simulation.parameters
+    iceflow_prob_remake = remake(iceflow_prob; p=container)
```

Comment:
So just to follow, one of the main changes to adapt this is to avoid using a closure and instead use a `remake` to pass the SciMLStruct with the simulation and the parameters?